applications/GeoMechanicsApplication/tests/cpp_tests/test_time_stepping.cpp

```diff
@@ -0,0 +1,154 @@
+    void Initialize()             override {mSolverStrategyInitializeCalls += 1;}
+    void InitializeSolutionStep() override {mSolverStrategyInitializeSolutionStepCalls += 1;}
+    void Predict()                override {mSolverStrategyPredictCalls += 1;}
+    bool SolveSolutionStep()      override {mSolverStrategySolveSolutionsStepCalls += 1;
+                                            return true;}
+    void FinalizeSolutionStep()   override {mSolverStrategyFinalizeSolutionStepCalls += 1;}
+    unsigned int NumberOfSolverStrategyInitializeCalls()             const {return mSolverStrategyInitializeCalls;}
+    unsigned int NumberOfSolverStrategyInitializeSolutionStepCalls() const {return mSolverStrategyInitializeSolutionStepCalls;}
+    unsigned int NumberOfSolverStrategyPredictCalls()                const {return mSolverStrategyPredictCalls;}
+    unsigned int NumberOfSolverStrategySolveSolutionStepCalls()      const {return mSolverStrategySolveSolutionsStepCalls;}
+    unsigned int NumberOfSolverStrategyFinalizeSolutionStepCalls()   const {return mSolverStrategyFinalizeSolutionStepCalls;}
+
+private:
+    TimeStepEndState::ConvergenceState mConvergenceNeeded;
+    unsigned int mSolverStrategyInitializeCalls             = 0;
+    unsigned int mSolverStrategyInitializeSolutionStepCalls = 0;
+    unsigned int mSolverStrategyPredictCalls                = 0;
+    unsigned int mSolverStrategySolveSolutionsStepCalls     = 0;
+    unsigned int mSolverStrategyFinalizeSolutionStepCalls   = 0;
+};
+
+}
+
+namespace Kratos::Testing
+{
+
+KRATOS_TEST_CASE_IN_SUITE(RunReturnsNonConvergedWhenStrategyDoesNotConverge, KratosGeoMechanicsFastSuite)
+{
+    TimeStepExecutor executor;
+    auto nonconverging_strategy = std::make_shared<DummyStrategyWrapper>(false);
+    executor.SetSolverStrategy(nonconverging_strategy);
+    const auto time = 0.0;
+    KRATOS_EXPECT_EQ(TimeStepEndState::ConvergenceState::non_converged, executor.Run(time).convergence_state);
+}
+
+KRATOS_TEST_CASE_IN_SUITE(RunReturnsConvergedWhenStrategyConverged, KratosGeoMechanicsFastSuite)
+{
+    TimeStepExecutor executor;
+    auto converging_strategy = std::make_shared<DummyStrategyWrapper>(true);
+    executor.SetSolverStrategy(converging_strategy);
+    const auto time = 0.0;
+    KRATOS_EXPECT_EQ(TimeStepEndState::ConvergenceState::converged, executor.Run(time).convergence_state);
+}
+
+KRATOS_TEST_CASE_IN_SUITE(ProcessMemberFunctionsAllCalledOnce, KratosGeoMechanicsFastSuite)
+{
+    TimeStepExecutor executor;
+    auto converging_strategy = std::make_shared<DummyStrategyWrapper>(true);
+    executor.SetSolverStrategy(converging_strategy);
+    ProcessSpy spy;
+    TimeStepExecutor::ProcessRefVec process_refs{spy};
+    executor.SetProcessReferences(process_refs);
+    const auto time = 0.0;
+
+    executor.Run(time);
+
+    KRATOS_EXPECT_EQ(1,spy.NumberOfExecuteInitializeSolutionStepCalls());
+    KRATOS_EXPECT_EQ(1,spy.NumberOfExecuteFinalizeSolutionStepCalls());
```

Comment:
Nit-picking from my side:
```suggestion
    KRATOS_EXPECT_EQ(1, spy.NumberOfExecuteInitializeSolutionStepCalls());
    KRATOS_EXPECT_EQ(1, spy.NumberOfExecuteFinalizeSolutionStepCalls());
```

Reply:
done with a smile